plugins/generator-1.17.1/forge-1.17.1/templates/biome.java.ftl

```diff
@@ -0,0 +1,455 @@
+            <#else>
+            mobSpawnInfo.addSpawn(${generator.map(spawnEntry.spawnType, "mobspawntypes")}, new MobSpawnSettings.SpawnerData(${entity.toString().replace(".CustomEntity", "")}.entity, ${spawnEntry.weight}, ${spawnEntry.minGroup}, ${spawnEntry.maxGroup}));
+            </#if>
+        	</#if>
+        </#list>
+
+        return new Biome.BiomeBuilder()
+            .precipitation(Biome.Precipitation.<#if (data.rainingPossibility > 0)><#if (data.temperature > 0.15)>RAIN<#else>SNOW</#if><#else>NONE</#if>)
+            .biomeCategory(Biome.BiomeCategory.${data.biomeCategory})
+            .depth(${data.baseHeight}f)
+            .scale(${data.heightVariation}f)
+            .temperature(${data.temperature}f)
+            .downfall(${data.rainingPossibility}f)
+            .specialEffects(effects)
+            .mobSpawnSettings(mobSpawnInfo.build())
+            .generationSettings(biomeGenerationSettings.build())
+            .build();
+    }
+
+	@SubscribeEvent public void init(FMLCommonSetupEvent event) {
+    <#if data.biomeDictionaryTypes?has_content>
+    	BiomeDictionary.addTypes(ResourceKey.create(Registry.BIOME_REGISTRY, BuiltinRegistries.BIOME.getKey(${JavaModName}Biomes.${registryname?upper_case})),
+    	<#list data.biomeDictionaryTypes as biomeDictionaryType>
+        BiomeDictionary.Type.${generator.map(biomeDictionaryType, "biomedictionarytypes")}<#if biomeDictionaryType?has_next>,</#if>
+    	</#list>
+    	);
+    </#if>
+    <#if data.spawnBiome>
+        BiomeManager.addBiome(BiomeManager.BiomeType.${data.biomeType},
+            new BiomeManager.BiomeEntry(ResourceKey.create(Registry.BIOME_REGISTRY, BuiltinRegistries.BIOME.getKey(${JavaModName}Biomes.${registryname?upper_case})), ${data.biomeWeight}));
+    </#if>
+	}
+
+	<#if (data.treeVines?has_content && !data.treeVines.isEmpty())>
+	private static class CustomLeaveVineTreeDecorator extends LeaveVineTreeDecorator {
+
+    public static final CustomLeaveVineTreeDecorator instance = new CustomLeaveVineTreeDecorator();
+    public static com.mojang.serialization.Codec<LeaveVineTreeDecorator> codec;
+    public static TreeDecoratorType tdt;
+
+    static {
+    	codec = com.mojang.serialization.Codec.unit(() -> instance);
+    	tdt = new TreeDecoratorType(codec);
+    	tdt.setRegistryName("${registryname}_lvtd");
+    	ForgeRegistries.TREE_DECORATOR_TYPES.register(tdt);
+    }
+
+    @Override protected TreeDecoratorType<?> func_230380_a_() {
+    	return tdt;
+    }
+
+    @Override protected void func_227424_a_(IWorldWriter ww, BlockPos bp, BooleanProperty bpr, Set<BlockPos> sbc, MutableBoundingBox mbb) {
+    	this.func_227423_a_(ww, bp, ${mappedBlockToBlockStateCode(data.treeVines)}, sbc, mbb);
+    }
+
+	}
+
+	private static class CustomTrunkVineTreeDecorator extends TrunkVineTreeDecorator {

```

Comment:
We should not make inner classes anymore, to make the code more human-like

Reply:
It's not "translated" in mojmap yet. I didn't touch it yet. It was the next thing I wanted to do.